cache4k/src/commonMain/kotlin/io/github/reactivecircus/cache4k/CacheEvent.kt

```diff
@@ -0,0 +1,41 @@
+package io.github.reactivecircus.cache4k
+
+/**
+ * An event resulting from a mutative [Cache] operation.
+ */
+public data class CacheEvent<Key : Any, Value : Any>(
+    val type: CacheEventType,
+    val key: Key,
+    val oldValue: Value?,
+    val newValue: Value?,
+)
+
+public enum class CacheEventType {
+    Created,
+    Updated,
+    Removed,
+    Expired,
+    Evicted,
+}
```

Comment:
I wonder if we can combine these and have something like:

```kt
public sealed interface CacheEvent<Key: Any, Value: Any> {
    class Created(val value: Value): CacheEvent
    class Updated(val oldValue: Value, val newValue: Value): CacheEvent
    ...
}
```
So we don't need any nullable `Value?`.

Also the [new library guideline](https://kotlinlang.org/docs/jvm-api-guidelines-backward-compatibility.html#don-t-use-data-classes-in-api) recommends not using data class in libraries for better backward compatibility support.

Reply:
So it's will be look like this:
```kotlin
public sealed interface CacheEvent<Key : Any, Value : Any> {
    public val key: Key

    public val oldValue: Value?
        get() = when (this) {
            is Created -> null
            is Updated -> oldValue
            is Evicted -> value
            is Expired -> value
            is Removed -> value
        }

    public class Created<Key : Any, Value : Any>(
        override val key: Key,
        public val value: Value,
    ) : CacheEvent<Key, Value>

    public class Updated<Key : Any, Value : Any>(
        override val key: Key,
        public override val oldValue: Value,
        public val newValue: Value,
    ) : CacheEvent<Key, Value>

    public class Removed<Key : Any, Value : Any>(
        override val key: Key,
        public val value: Value,
    ) : CacheEvent<Key, Value>

    public class Expired<Key : Any, Value : Any>(
        override val key: Key,
        public val value: Value,
    ) : CacheEvent<Key, Value>

    public class Evicted<Key : Any, Value : Any>(
        override val key: Key,
        public val value: Value,
    ) : CacheEvent<Key, Value>
}
```
I want add `oldValue` getter for case when we need close old resource:
```kotlin
val cache = Cache.Builder<Long, String>()
    .eventListener { event ->
        event.oldValue?.close()
    }
    .build()
```
Do you mind?

Reply:
I'm not sure what other events we might need in the future, `oldValue` doesn't seem to make sense for a generic `CacheEvent`. Maybe we should avoid adding this API for now.

Also re. `invalidateAll()`, does it make sense to add a `object RemovedAll`? instead of calling it with `Removed` for every item?

Reply:
With `object RemovedAll`, if we want to do something with values we need call `asMap()` before `invalidateAll`.

With `Removed` events, we can handle all values in one place.

```kotlin
val cache = Cache.Builder<Long, String>()
    .eventListener { event ->
        // do something
    }
    .build()
...
cache.asMap().entries.forEach { value ->
    // do something
}
cache.invalidateAll()
```
vs
```kotlin
val cache = Cache.Builder<Long, String>()
    .eventListener { event ->
        // do something
    }
    .build()
...
cache.invalidateAll()
```

Reply:
Also an event listener can be declared as a class and we can handle events in many caches. With `object RemovedAll` it will be difficult if we need to work with values.
```
class CustomEventListener : CacheEventListener<Long, String> {
    override fun onEvent(event: CacheEvent<Long, String>) {
        // do something
    }
}
val customEventListener = CustomEventListener()

val cache1 = Cache.Builder<Long, String>()
    .eventListener(customEventListener)
    .build()

val cache2 = Cache.Builder<Long, String>()
    .eventListener(customEventListener)
    .build()
```

Reply:
Sounds good 👍 I'll take  another look tomorrow